validator/core/src/output/path.rs

```diff
@@ -0,0 +1,202 @@
+//! Module for handling tree paths.
+//!
+//! The [`PathElement`], [`Path`], and [`PathBuf`] types are used to uniquely
+//! refer to any node in a Substrait plan (or, more accurately, any
+//! combination of protobuf and YAML data). [`Path`], and [`PathBuf`] work
+//! roughly the same as [`std::path::Path`], and [`std::path::PathBuf`], but
+//! for protobuf/YAML tree paths rather than filesystem paths.
+
+use crate::output::primitive_data;
+
+/// Element of a path to some field of a protobuf message and/or YAML file.
+#[derive(Clone, Debug, PartialEq)]
+pub enum PathElement {
+    /// Refers to an optional protobuf field with the given name within the
+    /// message, or a YAML map entry with the given key.
+    Field(String),
+
+    /// Refers to one of the elements of a repeated field with the given
+    /// name within the message referred to by the parent path.
+    Repeated(String, usize),
+
+    /// Refers to the selected variant of a OneOf field with the given name
+    /// within the message referred to by the parent path. The first str is
+    /// the field name, the second is the variant name.
+    Variant(String, String),
+
+    /// Refers to an indexed element within a YAML array.
+    Index(usize),
+}
+
+impl std::fmt::Display for PathElement {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if !f.alternate() {
+            match self {
+                PathElement::Index(_) => {}
+                _ => write!(f, ".")?,
+            }
+        }
+        match self {
+            PathElement::Field(field) => write!(f, "{}", primitive_data::as_ident_or_string(field)),
+            PathElement::Repeated(field, index) => write!(
+                f,
+                "{}[{}]",
+                primitive_data::as_ident_or_string(field),
+                index
+            ),
+            PathElement::Variant(field, variant) => write!(
+                f,
+                "{}<{}>",
+                primitive_data::as_ident_or_string(field),
+                primitive_data::as_ident_or_string(variant)
+            ),
+            PathElement::Index(index) => write!(f, "[{}]", index),
```

Comment:
Random observation that probably applies to other format strings, but you can also do this now:
```suggestion
            PathElement::Index(index) => write!(f, "[{index}]"),
```

Reply:
Cool, didn't know that was a thing.